src/goto-instrument/reachability_slicer.cpp

```diff
@@ -138,51 +240,81 @@ void reachability_slicert::fixedpoint_from_assertions(
+      // visit the call instruction's local successor now.
+      forward_walk_call_instruction(node, stack, called_function_heads);
+    }
+    else
+    {
+      // General case, including end of function: queue all successors.
+      for(const auto &edge : node.out)
+        stack.push_back(edge.first);
+    }
+  }
 
-        // Check if it can return, and if so mark the callsite's successor:
-        while(!callee_it->is_end_function())
-          ++callee_it;
+  return called_function_heads;
+}
 
-        if(!cfg[cfg.entry_map[callee_it]].out.empty())
-        {
-          stack.emplace_back(
-            cfg.entry_map[std::next(node.PC)], caller_is_known);
-        }
-      }
-      else
-      {
-        // Bodyless function -- mark the successor instruction only.
-        stack.emplace_back(successor_index, caller_is_known);
-      }
+/// Perform forwards depth-first search of the control-flow graph of the
+/// goto program, starting from a given set of nodes. Steps into callsites;
+/// ignores return sites, which have been taken care of by
+/// `forward_outwards_walk_from`. Note it is important this is done *after*
+/// the outwards walk, because the outwards walk visits strictly more functions
+/// as it visits all possible callers.
+/// \param stack: nodes to start from
+void reachability_slicert::forward_inwards_walk_from(
+  std::vector<cfgt::node_indext> stack)
+{
+  while(!stack.empty())
+  {
+    auto &node = cfg[stack.back()];
+    stack.pop_back();
+
+    if(node.reachable_from_assertion)
+      continue;
+    node.reachable_from_assertion = true;
+
+    if(node.PC->is_function_call())
+    {
+      // Visit both the called function head and the callsite successor:
+      forward_walk_call_instruction(node, stack, stack);
     }
-    else if(node.PC->is_end_function() && caller_is_known)
+    else if(node.PC->is_end_function())
     {
       // Special case -- the callsite successor was already queued when entering
       // this function, more precisely than we can see from the function return
       // edges (which lead to all possible callers), so nothing to do here.
     }
     else
     {
-      // General case, including end-of-function where we don't know our caller.
-      // Queue all successors.
+      // General case: queue all successors.
       for(const auto &edge : node.out)
-        stack.emplace_back(edge.first, caller_is_known);
+        stack.push_back(edge.first);
     }
   }
 }
 
+/// Perform forwards depth-first search of the control-flow graph of the
+/// goto program, starting from the nodes corresponding to the criterion and
+/// the instructions that might be executed concurrently. Set reaches_assertion
+/// to true for every instruction visited.
+/// \param is_threaded Instructions that might be executed concurrently
+/// \param criterion the criterion we are trying to hit
+void reachability_slicert::fixedpoint_from_assertions(
+  const is_threadedt &is_threaded,
+  slicing_criteriont &criterion)
```

Comment:
what are the side effects on this?

Reply:
None. Pushed an extra commit to constify these, at the risk of @tautschnig's ire re: scope creep ;)

Reply:
There you go :-)